x-pack/test/security_api_integration/tests/session_concurrent_limit/global_limit.ts

```diff
@@ -208,6 +213,8 @@ export default function ({ getService }: FtrProviderContext) {
         Array.from({ length: 10 }).map(() => loginWithBasic(testUser))
       );
 
+      await es.indices.refresh({ index: '.kibana_security_session*' });
```

Comment:
The index needs to be refreshed in order for the session limit to be enforced. If we don't use 'wait_for', it is unclear that we can properly enforce session limits for concurrent sessions that are created at the same time. Is this a real use case? Do we want to support it?

Reply:
> The index needs to be refreshed in order for the session limit to be enforced. If we don't use 'wait_for', it is unclear that we can properly enforce session limits for concurrent sessions that are created at the same time

That's correct, but we have a `1s` refresh interval configured for the index. That means that _in the majority_ of the cases there will be just a small window (up to a few seconds) where the limit could be bypassed. Even though the problem might be more apparent for highly loaded clusters, I'd still say it's an acceptable tradeoff, at least for the initial release.

On the other hand, it's a relatively small and low-risk change to use `refresh=wait_for` only when concurrent session limit is configured, but I'm wondering if it'd make sense to make it only when/if we learn that it's not what our customers and users are comfortable with. What do you think @jeramysoucy @legrego @arisonl?

Reply:
If we know of significant real life use cases where this will be a problem, then I would say let's default to using `wait_for` if concurrent session limits are configured. Otherwise, I like moving forward with `refresh=false` for now and potentially making `wait_for` a future option for flexibility should the need arise.

Reply:
> If we know of significant real life use cases where this will be a problem, then I would say let's default to using `wait_for` if concurrent session limits are configured. Otherwise, I like moving forward with `refresh=false` for now and potentially making `wait_for` a future option for flexibility should the need arise.

I'm comfortable with this, but I think we should call this out as a limitation in our docs for this feature. Otherwise, a security researcher could reasonably assert that this control is not working as intended.

Reply:
I would like to better understand the details of the underlying technical issue (we'll discuss offline), but in principle I agree that a documented small (a few seconds) deviation from the expected behavior is not an issue, if documented. However, there is this reference: "_in the majority_ of the cases there will be just a small window" which worries me a bit. Is there a chance that we our off by more than just a few seconds?

Reply:
> Is there a chance that we our off by more than just a few seconds?

I can think of two possible cases at the moment:
* If the ES cluster is overloaded, but in this case one cannot do much with the cluster anyway.
* If admins mistakenly applied another "refresh_interval" to the Kibana internal session index, but I'm not even sure if it's possible today with all the improvements ES did for the internal/hidden/system indices during the last two years.

Reply:
As discussed offline, we all agree that the best path is to proceed and document, since we doubt that any deviation more significant than a few seconds is even possible. If it comes up, we can reassess.

Reply:
> That's correct, but we have a `1s` refresh interval configured for the index. That means that _in the majority_ of the cases there will be just a small window (up to a few seconds) where the limit could be bypassed.

Just so I fully understand the problem. If we use `refresh=wait_for`, we are essentially synchronizing the response from the request _Persist session_ in the diagram below, correct?

<img width="851" alt="image" src="https://user-images.githubusercontent.com/5469006/221376776-d714b19b-5efe-49dc-99dc-d208fe970c15.png"> _Scenario when 2 users try to log in with the same creds almost concurrently (in the same refresh interval). Pink refers to Kibana Browser<->Server requests, and blue to Kibana<->ES. The requests to **Persist session** end at the refresh interval, and the data is available to search immediately after that. **Search for other sessions** finds 2 sessions._

When using `refresh=false`, the following may occur:

<img width="827" alt="image" src="https://user-images.githubusercontent.com/5469006/221376983-1a36dd98-c25f-4147-82aa-4205352a227e.png"> _The search occurs before the refresh interval has kicked in => **Search for other sessions** finds none of the sessions._

AFAIK, ES is looking at ways to make the data available for search even before the refresh interval has kicked in. Once that's complete, we may be able to completely migrate to `refresh=wait_for`. 🥳

Reply:
> Just so I fully understand the problem. If we use refresh=wait_for, we are essentially synchronizing the response from the request Persist session in the diagram below, correct?

Yeah, it's close! The only difference is that we don't search for other sessions during login, but whenever the session is retrieved in the subsequent requests to make sure the session limitation isn't bypassed with tons of concurrent logins. But the idea is the same.

> AFAIK, ES is looking at ways to make the data available for search even before the refresh interval has kicked in. Once that's complete, we may be able to completely migrate to refresh=wait_for. partying_face

Oh, that would be great, thanks for sharing! Is it something ES team has on the near-term roadmap (before Oct 2023)?

Reply:
I'm not sure of their current status on this, however the internal issue (ES-4076) seems to indicate May 2023.

Reply:
Regarding augmenting our documentation to cover the known limitations of this change, do we want to make the note where we document the concurrent limit configuration option? https://www.elastic.co/guide/en/kibana/master/xpack-security-session-management.html#session-max-sessions